File: src/main/java/com/portofino/realtrainmodunofficial/rail/util/RailMaker.java
package com.portofino.realtrainmodunofficial.rail.util;

import java.util.ArrayList;
import java.util.List;

public final class RailMaker {
    private final List<RailPosition> rpList;

    public RailMaker(List<RailPosition> positions) {
        this.rpList = new ArrayList<>(positions);
    }

    public RailMaker(RailPosition[] positions) {
        this.rpList = new ArrayList<>(positions.length);
        for (RailPosition position : positions) {
            if (position != null) {
                this.rpList.add(position);
            }
        }
    }

    private SwitchType getSwitchType() {
        if (this.rpList.size() == 3) {
            int count = 0;
            for (RailPosition rp : this.rpList) {
                count += rp.switchType == 1 ? 1 : 0;
            }
            if (count == 1) {
                return new SwitchType.SwitchBasic();
            }
        } else if (this.rpList.size() == 4) {
            int count = 0;
            for (RailPosition rp : this.rpList) {
                count += rp.switchType == 1 ? 1 : 0;
            }
            if (count == 2) {
                return new SwitchType.SwitchSingleCross();
            }
            if (count == 4) {
                for (int i = 0; i < this.rpList.size(); ++i) {
                    for (int j = i + 1; j < this.rpList.size(); ++j) {
                        if ((this.rpList.get(i).direction & 7) == (this.rpList.get(j).direction & 7)) {
                            return new SwitchType.SwitchScissorsCross();
                        }
                    }
                }
                return new SwitchType.SwitchDiamondCross();
            }
        }
        return null;
    }

    public SwitchType getSwitch() {
        SwitchType type = this.getSwitchType();
        if (type == null) {
            return null;
        }

        List<RailPosition> switchList = new ArrayList<>();
        List<RailPosition> normalList = new ArrayList<>();
        for (RailPosition rp : this.rpList) {
            if (rp.switchType == 1) {
                switchList.add(rp);
            } else {
                normalList.add(rp);
            }
        }
        return type.init(switchList, normalList) ? type : null;
    }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/rail/util/RailMaker.kt
// SPDX-License-Identifier: LGPL-3.0-or-later
// Copyright © 2026 mirukuneko and RealTrainModRenewed contributors
package cc.mirukuneko.realtrainmodrenewed.rail.util

import java.lang.reflect.Array as ReflectArray

class RailMaker private constructor(positions: List<RailPosition>, @Suppress("UNUSED_PARAMETER") collected: Unit) {
    private val rpList: MutableList<RailPosition>

    constructor(positions: List<RailPosition>) : this(positions, Unit)

    constructor(positions: Array<RailPosition?>) : this(
        positions.filterNotNull().toList(),
        Unit,
    )

    /**
     * Script compatibility constructor. Nashorn arrays implement Map rather than
     * Java List/Array, while other engines can expose Iterable or reflective arrays.
     */
    constructor(positions: Any?) : this(RailPositionContainers.collect(positions), Unit)

    /** Legacy RailMaker(world, positions, fixRTMRailMapVersion) signature. */
    constructor(
        @Suppress("UNUSED_PARAMETER") world: Any?,
        positions: Any?,
        @Suppress("UNUSED_PARAMETER") fixRTMRailMapVersion: Int,
    ) : this(positions)

    init {
        rpList = positions.toMutableList()
    }

    private fun getSwitchType(): SwitchType? {
        return when (rpList.size) {
            3 -> {
                val count = rpList.count { it.switchType == 1.toByte() }
                if (count == 1) SwitchType.SwitchBasic() else null
            }
            4 -> {
                val count = rpList.count { it.switchType == 1.toByte() }
                when {
                    count == 2 -> SwitchType.SwitchSingleCross()
                    count == 4 -> {
                        for (i in rpList.indices) {
                            for (j in i + 1 until rpList.size) {
                                if ((rpList[i].direction.toInt() and 7) == (rpList[j].direction.toInt() and 7)) {
                                    return SwitchType.SwitchScissorsCross()
                                }
                            }
                        }
                        SwitchType.SwitchDiamondCross()
                    }
                    else -> null
                }
            }
            else -> null
        }
    }

    fun getSwitch(): SwitchType? {
        val type = getSwitchType() ?: return null
        val switchList = rpList.filter { it.switchType == 1.toByte() }
        val normalList = rpList.filter { it.switchType != 1.toByte() }
        return if (type.init(switchList, normalList)) type else null
    }
}

internal object RailPositionContainers {
    private const val MAX_DEPTH = 4

    fun collect(source: Any?): List<RailPosition> {
        val result = ArrayList<RailPosition>()
        collectInto(source, result, 0)
        return result
    }

    private fun collectInto(source: Any?, result: MutableList<RailPosition>, depth: Int) {
        if (source == null || depth > MAX_DEPTH) return
        when (source) {
            is RailPosition -> result.add(source)
            is Map<*, *> -> source.values.forEach { collectInto(it, result, depth + 1) }
            is Iterable<*> -> source.forEach { collectInto(it, result, depth + 1) }
            else -> {
                if (!source.javaClass.isArray) return
                for (index in 0 until ReflectArray.getLength(source)) {
                    collectInto(ReflectArray.get(source, index), result, depth + 1)
                }
            }
        }
    }
}
